src/Raven.Server/Documents/Queries/GraphQueryOrderByFieldSorter.cs

```diff
@@ -0,0 +1,143 @@
+﻿using System;
+using System.Collections.Generic;
+using Raven.Client.Json;
+using Sparrow.Json;
+
+namespace Raven.Server.Documents.Queries
+{
+    internal class GraphQueryOrderByFieldSorter: IComparer<GraphQueryRunner.Match>
+    {
+        private OrderByField _field;
+        private string _alias;
+        private BlittablePath _path;
+
+        public GraphQueryOrderByFieldSorter(OrderByField field)
+        {
+            var fieldName = field.Name.Value;
+            var indexOfDot = fieldName.IndexOf('.');
+            if (indexOfDot < 0)
+                throw new NotSupportedException($"{GetType().Name} got an order by field: {fieldName} that isn't in the expected format of alias.fieldName");
+            _alias = fieldName.Substring(0, indexOfDot);
+            _path = new BlittablePath(fieldName.Substring(indexOfDot + 1, fieldName.Length - indexOfDot - 1));
+            _field = field;
+        }
+
+
+
+        public int Compare(GraphQueryRunner.Match x, GraphQueryRunner.Match y)
+        {
+            int order = _field.Ascending ? 1 : -1;
+            object xObject = null;
+            object yObject = null;
+            var xResult = x.GetResult(_alias);
+            var yResult = y.GetResult(_alias);
+            switch (xResult)
+            {
+                case Document xDocument:
+                    if (yResult is Document yDocument)
+                    {
+                        xDocument.EnsureMetadata();
+                        yDocument.EnsureMetadata();
+                        xObject = _path.Evaluate(xDocument.Data, true);
+                        yObject = _path.Evaluate(yDocument.Data, true);
+                        break;
+                    }
+                    ThrowMissmatchTypes(xResult, yResult);                   
+                    break;
+                case BlittableJsonReaderObject bjroX:
+                    if (yResult is BlittableJsonReaderObject bjroY)
+                    {
+                        xObject = _path.Evaluate(bjroX, true);
+                        yObject = _path.Evaluate(bjroY, true);
+                        break;
+                    }
+                    ThrowMissmatchTypes(xResult, yResult);
```

Comment:
What would be the default? How can we deal with that if types are different? Implicitly return `false`?

I'm for throwing an exception and be explicit about the reason if failed.

Reply:
The problem is that we don't have types, and a single bad document can cause the query to _fail_, which can take prod down.
We need to define the sort order and just let it go in that manner

Reply:
@ayende, in that case, we should maybe return -1 meaning the bad cases will always show up last I think we do the same with NULL_VALUE in v3.x
@arekpalinski there is no false here since it is a compare so it could be big small equal or throw
We could even return those values last regardless ordering so they won't show up when doing ascending or descending order.

Reply:
I'm fine with that

Reply:
i'm going to open a new PR it seems my commit from yesterday didn't get in